src/ts/utils/prosemirror.ts

```diff
@@ -122,45 +143,53 @@ export const getReplacementFragmentsFromReplacement = (
           currentPos: currentPos + patch.count
         };
       }
-      const $from = tr.doc.resolve(from + currentPos);
-      const $to = tr.doc.resolve(Math.min($from.pos + patch.count, tr.doc.nodeSize - 2));
+
+      const newFrom = from + currentPos;
+      const newTo = newFrom + patch.count;
 
       // If this patch removes chars, create a fragment for
       // the range, and leave the cursor where it is.
       if (patch.removed) {
+        const fragment = {
+          from: newFrom,
+          to: newTo,
+          type: "DELETE" as const
+        };
+
         return {
-          fragments: currentFragments.concat({
-            from: $from.pos,
-            to: $to.pos,
-            text: "",
-            marks: []
-          }),
+          fragments: currentFragments.concat(fragment),
           currentPos
         };
       }
 
       const prevFragment = currentFragments[currentFragments.length - 1];
-      const isThisPatchNew =
-        !!prevFragment && (prevFragment.to || 0) < $from.pos;
-
-      let marks;
-      if (isThisPatchNew) {
-        // If this patch adds characters and the previous patch left
-        // a range intact, inherit the marks from the last character
-        // of that range.
-        const $lastCharFrom = tr.doc.resolve($from.pos - 1);
-        marks = $lastCharFrom.marksAcross($from) || Mark.none;
+      const isInsertion = !!prevFragment && (prevFragment.to || 0) < newFrom;
+
+      let getMarks;
+      if (isInsertion) {
+        // If this patch is an insertion, inherit the marks from the last character
+        // of that patch's range. This ensures that text that expands upon an existing
+        // range shares the existing range's style.
+        getMarks = (localTr: Transaction) => {
+          const $newFrom = localTr.doc.resolve(newFrom)
+          const $lastCharFrom = localTr.doc.resolve(newFrom - 1);
```

Comment:
I was a bit confused at how this would work in all cases, until I realised the **b** in **eggs** -> **beggars** doesn't count as an `isInsertion`. Perhaps a small comment about this on line :166 would be helpful.

Reply:
Yeah, `isInsertion` doesn't cover exactly what's happening here – I've updated the name of the var, and the comment, to hopefully do a better job! 👍